runtime/compiler/runtime/ClassUnloadAssumption.cpp

```diff
@@ -580,8 +589,151 @@ int32_t TR_RuntimeAssumptionTable::countRatAssumptions()
    return count;
    }
 
+bool
+TR_RuntimeAssumptionTable::assumptionCanBeSerialized(TR_RuntimeAssumptionKind kind)
+   {
+   bool canBeSerialized;
+
+   switch (kind)
+      {
+      case RuntimeAssumptionOnClassUnload:
+      case RuntimeAssumptionOnRegisterNative:
+         {
+         canBeSerialized = true;
+         break;
+         }
+      default:
+         {
+         canBeSerialized = false;
+         break;
+         }
+      }
+
+   return canBeSerialized;
+   }
+
+uint8_t *
+TR_RuntimeAssumptionTable::serialize(J9JITConfig *jitConfig)
+   {
+   OMR::CriticalSection serializeAssumptions(assumptionTableMutex);
+
+   if (TR::Options::getCmdLineOptions()->getVerboseOption(TR_VerbosePerformance))
+      TR_VerboseLog::writeLineLocked(TR_Vlog_PERF, "Serializing Assumptions:");
+
+   uint8_t *bufferStart = (uint8_t *)TR_PersistentMemory::jitPersistentAlloc(getTotalSizeOfAssumptionsToBeSerialized() + sizeof(SerializedDataAtlas));
+   TR_ASSERT_FATAL(bufferStart, "Failed to allocate memory to serialize runtime assumptions\n");
+
+   SerializedDataAtlas *atlas = (SerializedDataAtlas *)bufferStart;
+   memset(atlas, 0, sizeof(SerializedDataAtlas));
+   atlas->_size = getTotalSizeOfAssumptionsToBeSerialized();
+
+   uint8_t *buffer = bufferStart;
+   buffer += sizeof(SerializedDataAtlas);
+
+   for (int k=0; k < LastAssumptionKind; k++) // for each table
+      {
+      TR_RatHT *hashTable = _tables + k;
+      size_t hashTableSize = hashTable->_spineArraySize;
+      for (size_t i = 0; i < hashTableSize; ++i) // for each bucket
+         {
+         OMR::RuntimeAssumption *cursor = hashTable->_htSpineArray[i];
+         while (cursor) // for each entry
+            {
+            if (!cursor->isMarkedForDetach())
+               {
+               TR::SentinelRuntimeAssumption *sentinel = (TR::SentinelRuntimeAssumption *)cursor->getSentinel();
+               TR_ASSERT_FATAL(sentinel, "Sentinel can't be NULL\n");
+
+               uint32_t size = cursor->size();
+               J9JITExceptionTable *owningMetadata = (J9JITExceptionTable *)sentinel->getOwningMetadata();
+               TR_ASSERT_FATAL(owningMetadata, "owningMetadata can't be NULL\n");
+
+               cursor->serialize(buffer, (uint8_t *)owningMetadata);
+
+               atlas->_numAssumptionsPerKind[k]._count++;
+               atlas->_numAssumptionsPerKind[k]._size += size;
+
+               owningMetadata->runtimeAssumptionList = NULL;
+               buffer += size;
+
+               OMR::RuntimeAssumption *next = cursor->getNext();
+
+               cursor->dequeueFromListOfAssumptionsForJittedBody();
+               TR_PersistentMemory::jitPersistentFree(cursor);
```

Comment:
If the plan is to continue running the JVM after taking a snpashot, this shouldn't be done.

Reply:
Considering https://github.com/eclipse/openj9/pull/11223 I updated the PR to not free the assumptions anymore.